src/java_bytecode/java_bytecode_instrument.cpp

```diff
@@ -0,0 +1,520 @@
+
+  side_effect_expr_throwt throw_expr;
+  throw_expr.add_source_location()=original_loc;
+  throw_expr.move_to_operands(exc);
+
+  code_ifthenelset if_code;
+  if_code.add_source_location()=original_loc;
+  if_code.cond()=cond;
+  if_code.then_case()=code_expressiont(throw_expr);
+
+  init_code.move_to_operands(if_code);
+
+  return init_code;
+}
+
+/// Checks whether the array access array_struct[idx] is out-of-bounds,
+/// and throws ArrayIndexOutofBoundsException/generates an assertion
+/// if necessary; Exceptions are thrown when the `throw_runtime_exceptions`
+/// flag is set. Otherwise, assertions are emitted.
+/// \par parameters: `array_struct`: the array being accessed
+/// `idx`: index into the array
+/// `original_loc: source location in the original code
+/// \return Based on the value of the flag `throw_runtime_exceptions`,
+/// it returns code that either throws an ArrayIndexOutPfBoundsException
+/// or emits an assertion checking the array access
+codet java_bytecode_instrumentt::check_array_access(
+  const exprt &array_struct,
+  const exprt &idx,
+  const source_locationt &original_loc)
+{
+  const constant_exprt &zero=from_integer(0, java_int_type());
+  const binary_relation_exprt ge_zero(idx, ID_ge, zero);
+  const member_exprt length_field(array_struct, "length", java_int_type());
+  const binary_relation_exprt lt_length(idx, ID_lt, length_field);
+  const and_exprt cond(ge_zero, lt_length);
+
+  if(throw_runtime_exceptions)
+    return throw_exception(
+      not_exprt(cond),
+      original_loc,
+      "java.lang.ArrayIndexOutOfBoundsException");
+
+  code_blockt bounds_checks;
+  bounds_checks.add(code_assertt(ge_zero));
+  bounds_checks.operands().back().add_source_location()=original_loc;
+  bounds_checks.operands().back().add_source_location()
+    .set_comment("Array index < 0");
+  bounds_checks.operands().back().add_source_location()
+    .set_property_class("array-index-out-of-bounds-low");
+
+  bounds_checks.add(code_assertt(lt_length));
+  bounds_checks.operands().back().add_source_location()=original_loc;
+  bounds_checks.operands().back().add_source_location()
+    .set_comment("Array index >= length");
+  bounds_checks.operands().back().add_source_location()
+    .set_property_class("array-index-out-of-bounds-high");
+
+  return bounds_checks;
+}
+
+/// Checks whether `class1` is an instance of `class2` and throws
+/// ClassCastException/generates an assertion when necessary;
+/// Exceptions are thrown when the `throw_runtime_exceptions` flag is set.
+/// Otherwise, assertions are emitted.
+/// \par parameters: `class1`: the subclass
+/// `class2`: the super class
+/// `original_loc: source location in the original code
+/// \return Based on the value of the flag `throw_runtime_exceptions`,
+/// it returns code that either throws an ClassCastException or emits an
+/// assertion checking the subtype relation
+codet java_bytecode_instrumentt::check_class_cast(
```

Comment:
Could we do with a testcase that casts to a derived type? For example, if C <: B <: A, something that casts an actual C from an A to a B (should pass but doesn't precisely match the target type)